pkg/check/datastore.go

```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
 	return c.RetrieveContent(tctx, []types.PbmProfileId{{UniqueId: name}})
 }
 
-func checkDataStore(dsName string, infrastructure *configv1.Infrastructure) error {
+func checkDataStore(ctx *CheckContext, dsName string, infrastructure *configv1.Infrastructure) error {
 	clusterID := infrastructure.Status.InfrastructureName
 	volumeName := fmt.Sprintf("[%s] 00000000-0000-0000-0000-000000000000/%s-dynamic-pvc-00000000-0000-0000-0000-000000000000.vmdk", dsName, clusterID)
 	klog.V(4).Infof("Checking data store %q with potential volume Name %s", dsName, volumeName)
 	if err := checkVolumeName(volumeName); err != nil {
 		return fmt.Errorf("datastore %s: %s", dsName, err)
 	}
+	if err := checkForDatastoreCluster(ctx, dsName); err != nil {
+		return err
+	}
 	return nil
 }
 
+func checkForDatastoreCluster(ctx *CheckContext, dataStoreName string) error {
+	tctx, cancel := context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	finder := find.NewFinder(ctx.VMClient, false)
+	datacenters, err := finder.DatacenterList(tctx, "*")
+	if err != nil {
+		klog.Errorf("error listing datacenters: %v", err)
+		return nil
+	}
+	workspaceDC := ctx.VMConfig.Workspace.Datacenter
+	var matchingDC *object.Datacenter
+	for _, dc := range datacenters {
+		if dc.Name() == workspaceDC {
+			matchingDC = dc
+		}
+	}
+
+	// lets fetch the datastore
+	finder = find.NewFinder(ctx.VMClient, false)
+	finder.SetDatacenter(matchingDC)
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	ds, err := finder.Datastore(tctx, dataStoreName)
+	if err != nil {
+		klog.Errorf("error getting datastore %s: %v", dataStoreName, err)
+		return nil
+	}
+
```

Comment:
This is used to get DC/DS elsewhere. Is it wrong?
https://github.com/openshift/vsphere-problem-detector/blob/ffe5e99596fb3c1a363c96ee057614bbaf997a15/pkg/check/folder.go#L22-L34

Reply:
the logic to fetch DS is same. I think I used somewhat different logic to fetch datacenter. I will make util function for that.

Reply:
What `checkForDatastoreCluster` / `CheckFolderPermissions` need is actually `getDatastore` (with `getDatacenter` inside) - fetching the datastore is copied over there.

Reply:
I refactored this code bit more. But in `checkForDatastoreCluster` function we are also fetching datastore summary/info so as we can get datastore URL. This field is not available in `object.Datastore` value.

Reply:
Thanks!